ccl/backupccl: TestBackupRestoreDataDriven failed

[(ccl/backupccl).TestBackupRestoreDataDriven failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2389870&tab=buildLog) on [master@33e3abc57cccedd86dd1b2e1d58feb075fcb8dd4](https://github.com/cockroachdb/cockroach/commits/33e3abc57cccedd86dd1b2e1d58feb075fcb8dd4):

```
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.(*restoreDataProcessor).Run()
      <autogenerated>:1 +0x57
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392 +0x432
  github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Run()
      <autogenerated>:1 +0x7f
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422 +0x94e
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.distRestore()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_processor_planning.go:228 +0x19a0
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.restore()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_job.go:621 +0xede
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.(*restoreResumer).Resume()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_job.go:1187 +0xbcb
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1089 +0x144
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1090 +0xa8e
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:245 +0x46a
  github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:853 +0x2c9
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x162

Goroutine 935 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x73
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:165 +0xda
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.runSplitAndScatter()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/split_and_scatter_processor.go:252 +0x2ef
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.(*splitAndScatterProcessor).Run.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/split_and_scatter_processor.go:182 +0x107

Goroutine 729 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:342 +0x14a
  github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:852 +0x49e
  github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:871 +0x16d
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.doRestorePlan()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_planning.go:1627 +0x27dd
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.restorePlanHook.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_planning.go:1297 +0x8bc
  github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/planhook.go:146 +0x10d
==================
FAIL	github.com/cockroachdb/cockroach/pkg/ccl/backupccl	48.900s

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestBackupRestoreDataDriven PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestBackupRestoreDataDriven.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/backupccl).TestBackupRestoreDataDriven failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2390505&tab=buildLog) on [master@75234294d22cc9a6e12724b043f49e78f0ac92d2](https://github.com/cockroachdb/cockroach/commits/75234294d22cc9a6e12724b043f49e78f0ac92d2):

```
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392 +0x432
  github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Run()
      <autogenerated>:1 +0x7f
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422 +0x94e
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.distRestore()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_processor_planning.go:228 +0x19a0
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.restore()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_job.go:621 +0xede
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.(*restoreResumer).Resume()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_job.go:1187 +0xbcb
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1089 +0x144
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1090 +0xa8e
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:245 +0x46a
  github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:853 +0x2c9
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x162

Goroutine 184 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x73
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:165 +0xda
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.runSplitAndScatter()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/split_and_scatter_processor.go:252 +0x2ef
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.(*splitAndScatterProcessor).Run.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/split_and_scatter_processor.go:182 +0x107

Goroutine 493 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:342 +0x14a
  github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:852 +0x49e
  github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:871 +0x16d
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.doRestorePlan()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_planning.go:1627 +0x27dd
  github.com/cockroachdb/cockroach/pkg/ccl/backupccl.restorePlanHook.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_planning.go:1297 +0x8bc
  github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/planhook.go:146 +0x10d
==================
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreDataDriven264680221
    --- FAIL: TestBackupRestoreDataDriven/user-defined-types (8.64s)
        testing.go:853: race detected during execution of test
    testing.go:853: race detected during execution of test

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestBackupRestoreDataDriven PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestBackupRestoreDataDriven.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
+++ b/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
@@ -197,6 +197,8 @@ func TestOracleFactory(t *testing.T) {
 func TestFollowerReadsWithStaleDescriptor(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
+	skip.WithIssue(t, 50032, "temporary")
+
 	// This test sleeps for a few sec.
 	skip.UnderShort(t)
 

--- a/pkg/util/tracing/tracer.go
+++ b/pkg/util/tracing/tracer.go
@@ -366,21 +366,6 @@ func (t *Tracer) startSpanGeneric(
 		},
 	}
 	s.crdb.mu.duration = -1 // unfinished
-	s.crdb.mu.tags = tags
-
-	// Copy baggage from parent.
-	//
-	// NB: this could be optimized.
-	for k, v := range parentContext.Baggage {
-		s.SetBaggageItem(k, v)
-	}
-
-	traceID := parentContext.TraceID
-	if traceID == 0 {
-		traceID = uint64(rand.Int63())
-	}
-	s.crdb.TraceID = traceID
-	s.crdb.SpanID = uint64(rand.Int63())
 
 	// We use the shadowTr from the parent context over that of our
 	// tracer because the tracer's might have changed and be incompatible.
@@ -400,6 +385,27 @@ func (t *Tracer) startSpanGeneric(
 			}
 		}
 	}
+
+	// Set initial tags.
+	//
+	// NB: this could be optimized.
+	for k, v := range tags {
+		s.SetTag(k, v)
+	}
+
+	// Copy baggage from parent.
+	//
+	// NB: this could be optimized.
+	for k, v := range parentContext.Baggage {
+		s.SetBaggageItem(k, v)
+	}
+
+	traceID := parentContext.TraceID
+	if traceID == 0 {
+		traceID = uint64(rand.Int63())
+	}
+	s.crdb.TraceID = traceID
+	s.crdb.SpanID = uint64(rand.Int63())
 
 	// Start recording if necessary.
 	if parentContext.recordingType != NoRecording {

--- a/pkg/util/tracing/tracer_span.go
+++ b/pkg/util/tracing/tracer_span.go
@@ -745,7 +745,7 @@ func IsBlackHoleSpan(s opentracing.Span) bool {
 // this is true, any Span derived from this context will be a "black hole Span".
 //
 // You should never need to care about this method. It is exported for technical
-// rasons.
+// reasons.
 func IsNoopContext(spanCtx opentracing.SpanContext) bool {
 	sc := spanCtx.(*SpanContext)
 	return sc.isNoop()
@@ -759,6 +759,9 @@ func (sc *SpanContext) isNoop() bool {
 // the Span tags.
 func SetSpanStats(os opentracing.Span, stats SpanStats) {
 	s := os.(*Span)
+	if s.isNoop() {
+		return
+	}
 	s.crdb.mu.Lock()
 	s.crdb.mu.stats = stats
 	for name, value := range stats.Stats() {
@@ -774,6 +777,9 @@ func (s *Span) Finish() {
 
 // FinishWithOptions is part of the opentracing.Span interface.
 func (s *Span) FinishWithOptions(opts opentracing.FinishOptions) {
+	if s.isNoop() {
+		return
+	}
 	finishTime := opts.FinishTime
 	if finishTime.IsZero() {
 		finishTime = time.Now()
@@ -831,6 +837,9 @@ func (s *Span) SpanContext() SpanContext {
 
 // SetOperationName is part of the opentracing.Span interface.
 func (s *Span) SetOperationName(operationName string) opentracing.Span {
+	if s.isNoop() {
+		return s
+	}
 	if s.ot.shadowSpan != nil {
 		s.ot.shadowSpan.SetOperationName(operationName)
 	}
@@ -840,6 +849,9 @@ func (s *Span) SetOperationName(operationName string) opentracing.Span {
 
 // SetTag is part of the opentracing.Span interface.
 func (s *Span) SetTag(key string, value interface{}) opentracing.Span {
+	if s.isNoop() {
+		return s
+	}
 	return s.setTagInner(key, value, false /* locked */)
 }
 
@@ -868,6 +880,9 @@ func (s *Span) setTagInner(key string, value interface{}, locked bool) opentraci
 
 // LogFields is part of the opentracing.Span interface.
 func (s *Span) LogFields(fields ...otlog.Field) {
+	if s.isNoop() {
+		return
+	}
 	if s.ot.shadowSpan != nil {
 		s.ot.shadowSpan.LogFields(fields...)
 	}
@@ -908,6 +923,9 @@ func (s *crdbSpan) LogFields(fields ...otlog.Field) {
 
 // LogKV is part of the opentracing.Span interface.
 func (s *Span) LogKV(alternatingKeyValues ...interface{}) {
+	if s.isNoop() {
+		return
+	}
 	fields, err := otlog.InterleavedKVToFields(alternatingKeyValues...)
 	if err != nil {
 		s.LogFields(otlog.Error(err), otlog.String("function", "LogKV"))
@@ -918,6 +936,9 @@ func (s *Span) LogKV(alternatingKeyValues ...interface{}) {
 
 // SetBaggageItem is part of the opentracing.Span interface.
 func (s *Span) SetBaggageItem(restrictedKey, value string) opentracing.Span {
+	if s.isNoop() {
+		return s
+	}
 	s.crdb.SetBaggageItemAndTag(restrictedKey, value)
 	if s.ot.shadowSpan != nil {
 		s.ot.shadowSpan.SetBaggageItem(restrictedKey, value)
@@ -971,11 +992,17 @@ func (s *Span) Tracer() opentracing.Tracer {
 
 // LogEvent is part of the opentracing.Span interface. Deprecated.
 func (s *Span) LogEvent(event string) {
+	if s.isNoop() {
+		return
+	}
 	s.LogFields(otlog.String(tracingpb.LogMessageField, event))
 }
 
 // LogEventWithPayload is part of the opentracing.Span interface. Deprecated.
 func (s *Span) LogEventWithPayload(event string, payload interface{}) {
+	if s.isNoop() {
+		return
+	}
 	s.LogFields(otlog.String(tracingpb.LogMessageField, event), otlog.Object("payload", payload))
 }
 

--- a/pkg/util/tracing/tracer_test.go
+++ b/pkg/util/tracing/tracer_test.go
@@ -277,7 +277,7 @@ func TestLightstepContext(t *testing.T) {
 		UseGRPC:        true,
 	})
 	tr.setShadowTracer(lightStepManager{}, lsTr)
-	s := tr.StartSpan("test")
+	s := tr.StartSpan("test").(*Span)
 
 	const testBaggageKey = "test-baggage"
 	const testBaggageVal = "test-val"
@@ -294,8 +294,8 @@ func TestLightstepContext(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	s2 := tr.StartSpan("child", opentracing.FollowsFrom(wireContext))
-	s2Ctx := s2.(*Span).ot.shadowSpan.Context()
+	s2 := tr.StartSpan("child", opentracing.FollowsFrom(wireContext)).(*Span)
+	s2Ctx := s2.ot.shadowSpan.Context()
 
 	// Verify that the baggage is correct in both the tracer context and in the
 	// lightstep context.

